### ISSUE TITLE
Testing the network on music datasets

### ISSUE THREAD

**Issue by ibab** [ibab/tensorflow-wavenet#104]
I've started to play around with the [MagnaTagATune dataset](http://mirg.city.ac.uk/codeapps/the-magnatagatune-dataset).
There's a small change that needs to be made to the code when training on this dataset:
Because it uses mp3 instead of wav, the pattern in [`wavenet/audio_reader.py`](wavenet/audio_reader.py) needs to be adjusted.
It would be nice to write a `MagnaReader` class that inherits from the `AudioReader` (or contains one), and that's able to filter the content by genre using the provided metadata.

**Comment by lemonzi** [comment]
I started training it on violin samples today as well, we'll see how it
goes! The AudioReader should be a bit more generic indeed, and then we can
either write wrapper for VCTK or Magna or subclasses.

El dc., 28 set. 2016 a les 16:44, Igor Babuschkin (<notifications@github.com>)
va escriure:

**Comment by Zeta36** [comment]
I've made a little experiment. I trained the WaveNet model using just a single wav file (4 seconds piano wav - 16bits - Mono). As I guessed, the loss dropped in hardly 4k steps to ~0.060, but when I generated the audio (using as parameter 64000 samples for 4 seconds generated), the result was a little strange. You can see that the model is playing someting like a piano sound (and the tempo looks pretty similar), but I have still a noisy background that I guessed it will not be there because the loss was just 0.060. It's possible this noise to be caused by the resting cross-entropy error in the prediction?

It's worth to note that I had to low the window of the sample in training time to 10.000 (SAMPLE_SIZE = 10000) because I have no GPU and I cannot wait 20 sec/step.

The other parameters are the default ones that are in the master json file.
I set the learning rate down every 1000 steps from 0.01 to 0.0001

This is the original wav file: [https://soundcloud.com/samu-283712554/piano-base](https://soundcloud.com/samu-283712554/piano-base)

And this is the progression of the training:
499 steps - https://soundcloud.com/samu-283712554/piano-499steps
1000 steps - https://soundcloud.com/samu-283712554/piano-1000
3000 steps - https://soundcloud.com/samu-283712554/piano-3000
3500 steps- https://soundcloud.com/samu-283712554/piano-3500

Could somebody with a good GPU repeat this experiment using the same base wav file (https://soundcloud.com/samu-283712554/piano-base)  but with a SAMPLE_SIZE of 100k and others parameters?

I don't know if I'm wrong but I think that if the model is right, this experiment should end quickly with a very smooth piano sound (without any noise), isn't it?

Well, if somebody finally do this in a while I'll be very pleased :).

Regards,
Samu.

**Comment by Zeta36** [comment]
I made another test with this. I generated the model but in this time a seeded the model with the original base wav file:

This is the result: https://soundcloud.com/samu-283712554/piano-3500-seed (the 4 first seconds are the original wav file seeded, and then there are the 4 seconds generated using the WaveNet trained -4k steps- model).

I don't know if it's normal but the first 4 seconds of the track (the 4 seconds primed from the original wav file), has too a little noisy background sound. I thought that part of the generate track should be clean and smooth, isn't it? is there some kind of problem with the librosa.load() method? or maybe with the trim_silence() one? 

I mean, when you seed with a wav in the generation process, this wav should be primed with the same quality in the generated wav, isn't it? or this loss of quality in the seed is due to the mu_law_encode and decode proccess?

Maybe the model doesn't train better becouse of this noisy background loaded by librosa and inserted in the input.

**Comment by ibab** [comment]
@Zeta36: I don't think there's an easy way to download your audio file from soundcloud.

**Comment by Zeta36** [comment]
I'm sorry, @ibab . Here you have it: [https://github.com/Zeta36/tensorflow-wavenet/blob/master/test/piano_base.wav](https://github.com/Zeta36/tensorflow-wavenet/blob/master/test/piano_base.wav)

**Comment by genekogan** [comment]
this is a simple change to `find_files` which looks for any audio files instead of just one set of extensions. i am experimenting with vocal tracks from music which are mixed. not sure there's any real need to limit to one extension since librosa handles pretty much any audio file.

```
def find_files(directory):
    '''Recursively finds aLl files matching the pattern.'''
    files = []
    for root, dirnames, filenames in os.walk(directory):
        for pattern in ['*.wav', '*.mp3', '*.aiff', '*.flac', '*.ape']:
            for filename in fnmatch.filter(filenames, pattern):
                files.append(os.path.join(root, filename))
    print("found %d files"%len(files))
    return files
```

**Comment by Zeta36** [comment]
@ibab , don't you think I should have get a clean and smooth wav after reaching a loss of only 0.060 in the test I commented above? Did you tried it by yourself?

**Comment by ibab** [comment]
@Zeta36: Can't use the GPUs at the moment, as they're needed for other things.
I don't think that a loss of 0.060 means that the generated samples will sound clean.
If the network highly overfits the sample, it could be very sensitive to even tiny differences in the input during generation.
That means that if one of the waveform samples it created deviates from what it received during training, the output will quickly leave the region of sample space that the network is familiar with.

If you look at #47, several people have posted very clean sounding samples, so it appears that this is something that gets better with more training samples and a substantially longer training period.

Also, the network configuration that you've used could be relevant to this. The results in #47 were achieved with substantially deeper networks than the one in the default `wavenet_params.json`.
Using 4 stacks of 1...512 dilation layers seems to be a good starting point.

Finally, there's also the possibility that things would are better when using the waveform directly as an input to the network instead of one-hot encoding it first, which is implemented in #106.

**Comment by Zeta36** [comment]
"That means that if one of the waveform samples it created deviates from what it received during training, the output will quickly leave the region of sample space that the network is familiar with."

Yes, that's true.

Thank you for the comments.

**Comment by maxhodak** [comment]
Here's a 5-second sample that is kind of not completely terrible: https://soundcloud.com/maxhodak/tycho-wavenet-version-1/s-YHbBX

trained to 20k steps, wav files of tycho's whole discography (individual tracks per file)

```
{
    "filter_width": 2,
    "sample_rate": 16000,
    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128,
                  1, 2, 4, 8, 16, 32, 64],
    "residual_channels": 32,
    "dilation_channels": 32,
    "quantization_channels": 256,
    "skip_channels": 512,
    "use_biases": true
}```

I tried training that out to 40k steps and it actually sounded a lot worse. Trying to understand why now.

**Comment by nakosung** [comment]
@maxhodak  I love Tycho! :)

**Comment by genekogan** [comment]
@maxhodak nice! the drums sound spot on...
what did you set the learning rate at?
i've also noticed that the loss does not fully translate to subjective quality. i have a few experiments i've been training, will post them here shortly.

**Comment by ibab** [comment]
> Trying to understand why subjectively it sounded much worse at 40k.
This could be a result of the network overfitting the dataset.
We can leave some files out of the training and monitor the loss on them to find out if that's the case.

**Comment by lemonzi** [comment]
Quick note -- for an online listening + easy download with no upload limit
cloud service, you can use Freesound rather than SoundCloud. Downside: it
doesn't allow the cool time-stamped comments.

El dt., 4 oct. 2016 a les 4:41, Igor Babuschkin (<notifications@github.com>)
va escriure:

**Comment by ibab** [comment]
@lemonzi: I wasn't aware of Freesound, thanks for pointing it out!

**Comment by robinsloan** [comment]
Just another music test to share… I think this is pretty nice/interesting!

https://soundcloud.com/robinsloan/sets/tensorflow-wavenet-temperature-demo

Model trained on [this album](https://skkakraba.bandcamp.com/album/songs-of-paapieye) to a loss of ~2.8 with these params:

```json
{
    "filter_width": 2,
    "sample_rate": 8000,
    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64],
    "residual_channels": 32,
    "dilation_channels": 32,
    "quantization_channels": 256,
    "skip_channels": 1024,
    "use_biases": false
}
```

**Comment by genekogan** [comment]
@robinsloan sounds very nice! i am also sharing some early results, though mine are much less coherent.

this is a sample generated on 2.5 hours of opera (Paisiello's Il Barbiere di Siviglia / Barber of Seville), with minimal pre-processing: https://soundcloud.com/genekogan/il-barbiere-di-siviglia-wavenet

this was generated after 25k steps with the following parameters and learning rate 0.001, loss around ~2.6.

```
{
    "filter_width": 2,
    "sample_rate": 16000,
    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512,
                  1, 2, 4, 8, 16, 32, 64],
    "residual_channels": 32,
    "dilation_channels": 32,
    "quantization_channels": 256,
    "skip_channels": 1024,
    "use_biases": true
}
```

it's mostly noise/gibberish, but to my ears, there is some coherent operatic material in there. i hear a sort of orchestra drone (sounds kind of like a warm-up/tuning) along with some faint baritone voices. right now it sounds more like granular synthesis than something coherent. i know i was too ambitious trying to train it on too many different sections, so i've started working on a script that uses librosa's analytical tools to narrow down a folder of a music to a more homogenous subset of it. i need to clean it up a bit and will post it as an ipython notebook shortly.

trying to get a sense of ways i can improve it. the obvious one i mentioned: limit to a smaller, more uniform set of audio. parametrically, having more dilations perhaps, lowering the sample rate, etc.

**Comment by robinsloan** [comment]
@genekogan, I like that a lot! There are some really interesting things happening in that clip. I mean it's almost like, forget simulation; on its own merits that's a novel, compelling sound. More, more!

Corpus selection is interesting. I went with SK Kakraba's gyil music because it's (a) stylistically pretty uniform, and (b) naturally quite noisy -- both of which seemed useful in this context.

**Comment by dunnevan** [comment]
I've been training on the MagnaTagATune dataset with clips that are tagged as solo piano.  

https://soundcloud.com/evan-dunn-676478257/sets/magnatagatune-solo-piano

I am training with batches of 4 with a loss fluctuating between ~1.7 to ~2.4. 

```
{
    "filter_width": 2,
    "sample_rate": 8192,
    "dilations": [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024,
                  1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024],
    "residual_channels": 32,
    "dilation_channels": 32,
    "quantization_channels": 256,
    "skip_channels": 1024,
    "use_biases": true,
    "scalar_input": false,
    "initial_filter_width": 32
}
```

**Comment by lmaxwell** [comment]
@dunnevan  nice!  What optimizer did you use? I noticed you run 400000 steps, have you adjusted learning rate during training?

**Comment by dunnevan** [comment]
@lmaxwell I'm using adam.  Started learning rate at 1e-3 and L2 at 1e-4 for the first 80000 steps then moved to 1e-4 and 1e-5.  I just recently pushed it down to another order of magnitude to see how it helps.  My feeling is that feeding it more data is the most important thing, even 400000 steps, 4 buckets at 8192 sampling is only about 3 minutes of trained audio.

**Comment by whuhyw** [comment]
@dunnevan  nice work! Sounds great! Can you describe your work work more specifically? i'd like to repeat you training.

**Comment by strongrex2001** [comment]
@dunnevan
![image](https://user-images.githubusercontent.com/23738781/33209660-e621a0b4-d151-11e7-9196-94a601113c32.png)
I tried with your parameters, however I didn't get satisfactory results... Is it because the different datasets of music should have different parameters...?

**Comment by pullahs** [comment]
Hey there,

I'm doing my masters degree on Sonic Arts and trying to come up with a good thesis on applications of machine learning on audio. I've been playing around with magenta and wavenet (thanks to you people!). Magenta is interesting as well but synthesizing sound from scratch, sample by sample is super interesting and revolutionary in my opinion. So, I decided to put my efforts on this one and I've been trying to coming up with some nice ideas lately.

So I'll try to explain some stuff that I did so far. 

**Training the model with Impulse Response files (IR) :**

I thought it would be a nice experiment to train the model with some IR (reverb impulse response) dataset and get the model synthesize some arbitrary spaces for me... So, I prepared a folder contains 115 IR samples of real spaces such as tunnel, hall, convention center etc.
I generated 20 second of IR sample after training the model up to 20000 steps by using the initial settings. This is what it gave me.
https://soundcloud.com/pullahs/generated-irs/s-WQW2v?in=pullahs/sets/wavenet/s-sx05a
which is quite interesting actually. (it worked...)
I sliced the IR wave file and extract some IRs out of it then placed it on a convolutional reverb plugin (that's how you use the IRs) to see the resulting space.
there is a straight forward playlist.
https://soundcloud.com/pullahs/sets/wavenet/s-sx05a

What I learned from this is that it promises a lot... With a better conditioning and more specific design approach on coding (which I'm super bad..) this can lead up to something much more useful. Designing auditory spaces...
The downside of my experiments is that this model only works in mono. To give a better perspective the model should work on stereo (at least). On "how could it be done (audiowise)" side of things, only thing that comes up to my mind is that while training the model it should also check the de-correlation between Left and Right channel. The phase difference between samples of Left channel and Right channel.

If the stereo implementation would be done or anyone would like to guide me on this one I have a lot  to try out.

Best.
